main/java/mjc/beans/AbstractPurchase.java

```diff
@@ -0,0 +1,76 @@
+package main.java.mjc.beans;
+
+import main.java.mjc.exceptions.CsvLineException;
+
+public abstract class AbstractPurchase implements Comparable<AbstractPurchase>, Cloneable {
+    private final Product product;
+    private final int amount;
+
+    public AbstractPurchase(String[] fields) throws Exception {
+        this(getValidPurchase(fields));
+    }
+
+    public AbstractPurchase(AbstractPurchase purchase) {
+        this(purchase.product, purchase.amount);
+    }
+
+    public AbstractPurchase(Product product, int amount) {
+        this.product = product;
+        this.amount = amount;
+    }
+
+    public AbstractPurchase(String name, int price, int amount) {
+        this.product = new Product(name, price);
+        this.amount = amount;
+    }
+
+    protected abstract Euro getFinalCost(Euro baseCost);
+
+    public Euro getCost() {
+        Euro baseCost = product.getPrice().mul(amount);
+        Euro finalCost = getFinalCost(baseCost).round(RoundMethod.FLOOR, 2);
+        return finalCost;
+    }
+
+    @Override
+    public String toString() {
+        return fieldsToString() + ';' + getCost();
+    }
+
+    protected String fieldsToString() {
+        return product + ";" + amount;
+    }
+
+    @Override
+    public int compareTo(AbstractPurchase purchase) {
+        return this.getCost().compareTo(purchase.getCost());
```

Comment:
i'd rather change the direction of sorting here than in `Euro`. 
```suggestion
        return -getCost().compareTo(purchase.getCost());
```